pkg/diagnose/connectivity/core_endpoint.go

```diff
@@ -68,112 +68,192 @@ func getLogsUseTCP() bool {
+		connType = "TCP"
+		url, err = logstcp.CheckConnectivityDiagnose(endpoints.Main, 5)
+	} else {
+		url, err = logshttp.CheckConnectivityDiagnose(endpoints.Main, pkgconfigsetup.Datadog())
+	}
+
+	name := fmt.Sprintf("%s connectivity to %s", connType, url)
+	return createDiagnosis(name, url, "", err)
+}
+
+func (cd *connDiagnostician) checkEndpointURL(url string, endpointInfo endpointInfo, apiKey string) diagnose.Diagnosis {
+	var responseBody []byte
+	var err error
+	var statusCode int
+	var httpTraces []string
+
+	if endpointInfo.Method == "HEAD" {
+		statusCode, err = sendHTTPHEADRequestToEndpoint(url, getClient(pkgconfigsetup.Datadog(), 1, cd.log, withOneRedirect()))
+	} else {
+		httpTraces = []string{}
+		ctx := httptrace.WithClientTrace(context.Background(), createDiagnoseTraces(&httpTraces, false))
+
+		statusCode, responseBody, _, err = sendHTTPRequestToEndpoint(ctx, cd.client, url, endpointInfo, apiKey)
+	}
+
+	// Check if there is a response and if it's valid
+	report, reportErr := verifyEndpointResponse(cd.diagCfg, statusCode, responseBody, err)
+
+	diagnosisName := "Connectivity to " + url
+	diag := createDiagnosis(diagnosisName, url, report, reportErr)
+
+	// Prepend http trace on error or if in verbose mode
+	if len(httpTraces) > 0 && (cd.diagCfg.Verbose || reportErr != nil) {
+		diag.Diagnosis = fmt.Sprintf("\n%s\n%s", strings.Join(httpTraces, "\n"), diag.Diagnosis)
+	}
+	return diag
+}
+
+func (cd *connDiagnostician) checkEndpoint(domainResolver resolver.DomainResolver, endpointInfo endpointInfo, apiKey string) diagnose.Diagnosis {
+	var url string
+
+	if endpointInfo.Endpoint.Name == "flare" {
+		url = endpointInfo.Endpoint.Route
+	} else {
+		domain := domainResolver.Resolve(endpointInfo.Endpoint)
+		url = createEndpointURL(domain, endpointInfo)
+	}
+	diag := cd.checkEndpointURL(url, endpointInfo, apiKey)
+
+	// Detect if the connection may have failed because a FQDN was used, by checking if one with a PQDN succeeds
+	if diag.Status != diagnose.DiagnosisSuccess && pkgconfigsetup.Datadog().Get("convert_dd_site_fqdn.enabled") == true && URLhasFQDN(url) {
+		pqdnURL := URLwithPQDN(url)
+		cd.log.Infof("The connection to %q with a FQDN failed; attempting to connect to %q", url, pqdnURL)
+
+		d := cd.checkEndpointURL(pqdnURL, endpointInfo, apiKey)
+		if d.Status == diagnose.DiagnosisSuccess {
+			diag.Remediation = fmt.Sprintf(
+				"The connection to the fully qualified domain name (FQDN) %q failed, but the connection to %q (without trailing dot) succeeded. Update your firewall and/or proxy configuration to accept FQDN connections, or disable FQDN usage by setting `convert_dd_site_fqdn.enabled` to false in the agent configuration.",
+				url, pqdnURL)
+		}
+	}
+	return diag
+}
+
+func URLhasFQDN(u string) bool {
+	url, err := url.Parse(u)
+	return err == nil && strings.HasSuffix(url.Hostname(), ".")
+}
+
+func URLwithPQDN(u string) string {
+	url, err := url.Parse(u)
+	if err != nil {
+		panic("Route is not a valid URL")
```

Comment:
This shouldn't be panic'ing. Could you have this method return an error instead?